File: sdk/go/cache.go
package experimentation

import (
	"container/list"
	"sync"
	"time"
)

// cacheEntry holds a single cached value with its expiry time and LRU list element.
type cacheEntry struct {
	key       string
	value     interface{}
	expiresAt time.Time
	element   *list.Element
}

// Cache is a thread-safe LRU cache with per-entry TTL expiry.
//
// When the cache is full, the least recently used entry is evicted to make room
// for the new entry. Expired entries are lazily removed on access.
type Cache struct {
	mu      sync.Mutex
	maxSize int
	ttl     time.Duration
	items   map[string]*cacheEntry
	lruList *list.List
}

// NewCache creates a new Cache with the given capacity and TTL.
// If maxSize <= 0, it defaults to 1. If ttl <= 0, entries never expire.
func NewCache(maxSize int, ttl time.Duration) *Cache {
	if maxSize <= 0 {
		maxSize = 1
	}
	return &Cache{
		maxSize: maxSize,
		ttl:     ttl,
		items:   make(map[string]*cacheEntry, maxSize),
		lruList: list.New(),
	}
}

// Get returns the cached value for key and true if found and not expired.
// If the entry has expired it is removed and (nil, false) is returned.
func (c *Cache) Get(key string) (interface{}, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()

	entry, ok := c.items[key]
	if !ok {
		return nil, false
	}

	// Check TTL expiry (ttl == 0 means no expiry).
	if c.ttl > 0 && time.Now().After(entry.expiresAt) {
		c.removeEntry(entry)
		return nil, false
	}

	// Move to front (most recently used).
	c.lruList.MoveToFront(entry.element)
	return entry.value, true
}

// Set stores a value in the cache under key.
// If the cache is full, the LRU entry is evicted first.
// If key already exists, its value and TTL are updated and it becomes MRU.
func (c *Cache) Set(key string, value interface{}) {
	c.mu.Lock()
	defer c.mu.Unlock()

	// Update existing entry.
	if entry, ok := c.items[key]; ok {
		entry.value = value
		if c.ttl > 0 {
			entry.expiresAt = time.Now().Add(c.ttl)
		}
		c.lruList.MoveToFront(entry.element)
		return
	}

	// Evict LRU if at capacity.
	if len(c.items) >= c.maxSize {
		c.evict()
	}

	// Insert new entry at front.
	var expiresAt time.Time
	if c.ttl > 0 {
		expiresAt = time.Now().Add(c.ttl)
	}
	entry := &cacheEntry{
		key:       key,
		value:     value,
		expiresAt: expiresAt,
	}
	entry.element = c.lruList.PushFront(entry)
	c.items[key] = entry
}

// Delete removes the entry for key from the cache, if it exists.
func (c *Cache) Delete(key string) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if entry, ok := c.items[key]; ok {
		c.removeEntry(entry)
	}
}

// Clear removes all entries from the cache.
func (c *Cache) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.items = make(map[string]*cacheEntry, c.maxSize)
	c.lruList.Init()
}

// Len returns the current number of entries in the cache (including expired ones
// that have not been lazily cleaned up yet).
func (c *Cache) Len() int {
	c.mu.Lock()
	defer c.mu.Unlock()

	return len(c.items)
}

// evict removes the least recently used item from the cache.
// Must be called with c.mu held.
func (c *Cache) evict() {
	back := c.lruList.Back()
	if back == nil {
		return
	}
	entry := back.Value.(*cacheEntry)
	c.removeEntry(entry)
}

// removeEntry deletes an entry from both the map and the LRU list.
// Must be called with c.mu held.
func (c *Cache) removeEntry(entry *cacheEntry) {
	c.lruList.Remove(entry.element)
	delete(c.items, entry.key)
}

File: sdk/js/src/cache.ts
interface CacheEntry<T> {
  value: T;
  expiresAt: number;
}

/** Per-user, per-key TTL cache. Only successful results are ever stored. */
export class UserKeyCache<T> {
  private readonly users = new Map<string, Map<string, CacheEntry<T>>>();

  constructor(private readonly ttlMs: number) {}

  get(userId: string, key: string): T | null {
    const entry = this.users.get(userId)?.get(key);
    if (!entry) return null;
    if (Date.now() > entry.expiresAt) {
      this.users.get(userId)?.delete(key);
      return null;
    }
    return entry.value;
  }

  set(userId: string, key: string, value: T): void {
    let byKey = this.users.get(userId);
    if (!byKey) {
      byKey = new Map();
      this.users.set(userId, byKey);
    }
    byKey.set(key, { value, expiresAt: Date.now() + this.ttlMs });
  }

  /** All live (non-expired) entries for a user, in insertion order. */
  entries(userId: string): Array<[string, T]> {
    const byKey = this.users.get(userId);
    if (!byKey) return [];
    const now = Date.now();
    const live: Array<[string, T]> = [];
    for (const [key, entry] of byKey) {
      if (now > entry.expiresAt) byKey.delete(key);
      else live.push([key, entry.value]);
    }
    return live;
  }

  clear(): void {
    this.users.clear();
  }
}
